rfc/002-nonzero-genesis.md

```diff
@@ -0,0 +1,102 @@
+it confusing that a given height can now refer to distinct states depending on the chain
+version.
+
+An ideal solution would be to always retain block backwards compatibility in such a way that chain 
+history is never lost on upgrades. However, this may require a significant amount of engineering
+work that is not viable for the planned Stargate release (Tendermint 0.34), and may prove too
+restrictive for future development.
+
+As a first step, allowing the new chain to start from an initial height specified in the genesis
+file would at least provide monotonically increasing heights. Information about the previous chain
+(ID and last block header) is also included, for informational purposes.
+
+External tooling would be required to map historical heights onto e.g. archive nodes that contain 
+blocks from previous chain version. Tendermint will not include any such functionality.
+
+## Proposal
+
+Tendermint will allow chains to start from an arbitrary initial height, with the minimum amount of
+changes to existing code and protocols:
+
+* A new field `initial_height` is added to the genesis file, defaulting to `1`. It can be set to any
+non-negative integer, and `0` is considered equivalent to `1`.
+
+* A new field `InitialHeight` is added to the ABCI `RequestInitChain` message, with the same value 
+and semantics as the genesis field.
+
+Additionally, the genesis file will be extended with a new `predecessor` field containing 
+information about the previous chain, specifically the final block header:
+
+```json
+{
+    "predecessor": {
+        "version": {"block": "10", "app": "0"},
+        "chain_id": "chain-1",
+        "height": 1000000,
+        "time": "2020-07-28T15:42:48.000Z",
+        "last_block_id": {
+            "hash": "2D7F34765B312A46BC551F9B3E0535D91CBA9513AAFDBB3458D17D7FD89FBEF0",
+            "parts": {"total": "1", "hash": "4D347E9C22E4C7EA15DB0B5AFF6799BC0A85B854443C3BA6E3B81A7BB7163931"},
+        },
+        "app_hash": "123AC50B458A998EB481D7E17A423C8C0ED238813518B55FAF57510F73F76DFC",
+        ...
+    }
+}
+```
+
+This field will be ignored by Tendermint, but may be helpful to integrators and users.
+
+If possible, no further changes will be made to any data structures. In particular, the initial
+height will not be added to the node state, and will instead be passed explicitly to any logic that 
+relies on it.
+
+ABCI applications may have to be updated to handle arbitrary initial heights, otherwise the initial
+block may fail.
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+* Heights will be unique throughout the history of a "logical" chain, across hard fork upgrades.
+
+### Negative
+
+* Upgrades still cause loss of block history.
+
+* Integrators will have to map height ranges to specific archive nodes/networks to query history.
+
+### Neutral
+
+* The predecessor block header is not verified by Tendermint at all, and must be verified by
```

Comment:
How would network governors verify it? I.e., imagine Alice has genesisA.json, Bob has genesisB.json. Since we do not hash a predecessor field, both Alice and Bob can have different predecessor blocks, and there's no way for me to find what is the "right" one. Then what's the reason for putting it in genesis in the first place.

Reply:
Oh, that's a good point. I somehow assumed that the entire genesis file was hashed into the first block. Does this field even have any value if it can't be trusted or agreed upon?

Reply:
no. I too thought we hash entire genesis into the first block or something like that a while ago, but no 😄

Reply:
I recall there was a proposal to do this long ago, but it never happened it seems.

Reply:
Ok, in that case I think we should just remove it. What's your take @ValarDragon?

Reply:
How is the Tendermint blockhash initialized? I would have assumed "LastBlockHash" for the first block in a chain is initialized to `Hash(genesis)`. Seems quite important for things like ensuring you agree on initial consensus params, and that not ensuring all nodes agree on all of the genesis would be a bug. 

For generic hashchains, you run into security proof problems under adaptively chosen inputs unless you do something like this. (Its possible its not a problem for Tendermint, but I'd be a bit surprised)

Reply:
`LastBlockID` and `LastCommitHash` are set to zero values for the initial block. All genesis fields (e.g. consensus params, validator sets, and the app hash) are hashed and/or handled individually in the genesis block, seems like the only field which isn't covered in some way is `GenesisTime`.

https://github.com/tendermint/tendermint/issues/4113 added a `--genesis_hash` parameter to verify the genesis file hash on startup. We could reuse e.g. `LastBlockID.Hash` or `LastCommitHash` for the genesis hash as well, but this might get confusing, especially with non-zero genesis or predecessor chains.

Reply:
Isn't all the data in the genesis effectively hashed in to the header already? the header of the first block i think should use the genesis time since there's otherwise no commit to draw it from, and everything else should be hashed in two (consensus params, validators, app state).

we don't take a literal hash of the json file but we should be able to verify one. though I guess that requires computing the application's merkle state from the app_state_bytes in the genessis, which might be prohibitive if you're just trying to verify that a genesis file is correct.

Not sure the best way to do it but possible the first LastBlockID should include the hash of the header or maybe we introduce a block-0 that has the genesis as the first transaction or something

Reply:
Huh I think the AppHash may not be currently making it into the first block because abci.InitChain does not return the app hash. It probably should?

Reply:
> Isn't all the data in the genesis effectively hashed in to the header already?

Yes, mostly. This appears sufficient to secure the current genesis file. However, we're considering adding a new `predecessor` field, but that field then won't be hashed into any blocks, and thus cannot be trusted.

Since it cannot be trusted it could be a potential social engineering attack vector, and I think we should err on the side of caution and drop this field until we find a way to secure the entire genesis file. The field is not actually used for anything anyway, so it's not like we _need_ it.

> maybe we introduce a block-0 that has the genesis as the first transaction or something

Would it be horrible to add a block field that's only used in the initial block?

> the AppHash may not be currently making it into the first block because abci.InitChain does not return the app hash

Seems so. That probably means that the genesis file app hash isn't verified either. I opened https://github.com/tendermint/tendermint/issues/5177 for this, we'll see if we can squeeze this in for 0.34 (the RC is already out).

Reply:
So should we open a separate issue in Tendermint to anchor the blockhash with the entire genesis file, and then do weak chain linking?

Reply:
Yeah, I've added https://github.com/tendermint/tendermint/issues/5183